### PR TITLE
Fix issue #5 and add ConfigManager

### PR DESCRIPTION
**Pull request by derpbuffalo**
As this is my first contribution, please let me know if I missed any contribution requirements. Happy to adjust and contribute.

### CODE REVIEW

**Comment by gui11aume** [comment]
@derpbuffalo This is amazing! Thanks for taking the time to help!

**Comment by gui11aume** [comment]
@derpbuffalo Would you be comfortable writing some test cases for the new functionalities? The changes look good already. We will accept them when we have some tests showing that they behave as expected. Let me know if you need some guidance. Otherwise, we can also take care of the tests but that will take some time.

**Comment by derpbuffalo** [comment]
Hey, sure I can write the tests as well, happy to do it next week!

**Comment by Adibvafa** [comment]
@derpbuffalo Looks great! A few question/suggestions:
- How can a user specify what type of mapping they preffer (or even no mapping?)
- ConfigManager could be a generic class, and a more specific ProteinConfig could inherit from it and be defined. This way we can have protein or even predict dna processing to inherit from it.

**Comment by gui11aume** [comment]
@derpbuffalo I agree with @Adibvafa, this is remarkable work. Thanks so much for your contribution! 👍

**Comment by derpbuffalo** [comment]
@Adibvafa 
I made ConfigManager and abstract class and added ProteinConfig.

Regarding how a user can specify the mapping:
The mapping type is always a dictionary of lists, the user can achieve following behaviors:
- having `ambiguous_aminoacid_behavior` set to `raise_error` (default) in the `ProteinConfig` is like having no mapping
- having `ambiguous_aminoacid_behavior` set to `standardize_deterministic` will always select the first item from the corresponding list in the mapping. If the user wishes, they can override the mapping with `ambiguous_aminoacid_map_override` to achieve a different result
- having `ambiguous_aminoacid_behavior` set to `standardize_random` will select a random item from the corresponding list in the mapping. Again, the user can override the list with `ambiguous_aminoacid_map_override`

Does this answer your question?

File: CodonTransformer/CodonUtils.py [hunk 1]
```diff
@@ -545,6 +546,132 @@ def __init__(self, data_path: str, train: bool = True, **kwargs):
...
+                if not isinstance(aminoacids, list):
+                    raise TypeError(f"Invalid type for aminoacids: {aminoacids}.")
+                if not aminoacids:
+                    raise ValueError(f"Override for aminoacid '{ambiguous_aminoacid}' cannot be empty list.")
+                if ambiguous_aminoacid not in AMBIGUOUS_AMINOACID_MAP:
+                    raise ValueError(f"Invalid amino acid in ambiguous_aminoacid_map_override: {ambiguous_aminoacid}")
+        else:
+            raise ValueError(f"Invalid configuration key: {key}")
+
+    def reset_config(self) -> None:
```
Comment (Adibvafa) [inline CodonTransformer/CodonUtils.py:665]: Could you change the default behaviour to `standardize_random`.
I will add an explanation in README describing the new functionality to change this default behvaiour.

**Review by Adibvafa — APPROVED** [review]
Amazing job on the PR! Once the default behaviour is changed, I will merge it.

File: CodonTransformer/CodonUtils.py [hunk 2]
```diff
@@ -545,6 +546,132 @@ def __init__(self, data_path: str, train: bool = True, **kwargs):
...
+        """
+        Reset the configuration to the default values.
+        """
+        self._config = {
+            'ambiguous_aminoacid_behavior': 'raise_error',
+            'ambiguous_aminoacid_map_override': {}
+        }
```
Comment (Adibvafa) [inline CodonTransformer/CodonUtils.py:672]: Could you change the default behaviour to standardize_random.
I will add an explanation in README describing the new functionality to change this default behvaiour.

**Comment by derpbuffalo** [comment]
@Adibvafa I changed the default behavior.

Thank you so much for both of you for your guidance!

**Comment by Adibvafa** [comment]
@derpbuffalo Great job with the PR!
Some tests seem to have failed, could you run it locally (using `pytest`) and ensure all test pass?

**Comment by derpbuffalo** [comment]
@Adibvafa My bad, I have built in a bug before the final commit, I fixed it and run the tests:

image

**Review by Adibvafa — APPROVED** [review]

**Comment by Adibvafa** [comment]
@derpbuffalo Great job! I will merge the PR. Thank you for your contribution 😄

**Pull request merged**